BUG: ensure that casting to/from structured is properly checked.

mhvk opened the pull request:
As noted in #11114, currently the following clearly wrong things happen:
```
np.can_cast(1., 'i2,i2', casting='safe')
# True
# while
np.can_cast(1., 'i2', casting='safe')
# False
np.array(1000.).astype('i1,i1', casting='safe')
# array((-24, -24), dtype=[('f0', 'i1'), ('f1', 'i1')])
# while
np.array(1000.).astype('i1', casting='safe')
# TypeError: Cannot cast array from dtype('float64') to dtype('int8') according to the rule 'safe'
```
The present PR fixes it, and has some elementary tests. In particular, it allows casting from simple to structured (EDIT) in unsafe mode, and from structured to simple only in unsafe mode *and* when the structured dtype has only a single field.

fixes #11114

p.s. Replaces #13646, which confused the general void type with structured.

mhvk committed:
BUG: ensure that casting to/from structured is properly checked.

Allow unsafe casting from a simple data type to a structured one with
multiple fields, but only from structured data type with a single
field (checked recursively) to a simple data type.

For now, continue to allow any structured to structured with
casting="unsafe", as the current can_cast implementation does not
match that of "astype".

eric-wieser commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -795,8 +811,33 @@ PyArray_CanCastTypeTo(PyArray_Descr *from, PyArray_Descr *to,
+            }
+            else {
+                return 0;
+            }
```
I could perhaps see an argument that this should be `any(PyArray_CanCastTypeTo(from, field, casting) for field in fields)`, allowing `2` to cast into `(2, 2, 2,)`, which I think we have precedent for

mhvk replied:
I was perhaps more driven by symmetry here than anything else. But I think you are right: casting to multiple fields is in principle not ambiguous. But it should be `all(cancast(...))`, then, not `any(...)`, correct?

seberg replied:
To take a step back here, do we want safe casting? It feels right, but it also feels right that safe casting in most cases aligns with "`common_type`" and I am not sure I feel that I feel things like `np.concatenate` should feel they can just "broadcast" to all fields without some manual casting. (I did not make up my mind yet, probably its OK).

eric-wieser replied:
To me it's a question of if we already set a precedent for this weird behavior. If we didn't, then the original more conservative patch was probably the right one.

mhvk replied:
I'm moving this out of the in-line comments, so it doesn't get lost when I change things.

mhvk commented:
@eric-wieser - I changed it to allow casting to multiple fields. Also realized I had to put it further up, since the changes hold also for unsafe casting. Since I think this is getting close, I also updated the docstrings. I think `can_cast` is now consistent with actually executing `a.astype(...)`.

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -793,6 +800,43 @@ PyArray_CanCastTypeTo(PyArray_Descr *from, PyArray_Descr *to,
+            PyDict_Next(from->fields, &ppos, NULL, &tuple);
+            return PyArray_CanCastTypeTo(
+                (PyArray_Descr *)PyTuple_GET_ITEM(tuple, 0),
+                to, casting);
```
OK, I think I disagree here, this is not safe casting, but only valid for unsafe casting.

seberg reviewed (commented):
I suppose a few tests assume that unsafe casting works in strange ways... I am not sure we want to go there and change that just now...

seberg reviewed (commented):
I suppose a few tests assume that unsafe casting works in strange ways... I am not sure we want to go there and change that just now...

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -2690,10 +2690,24 @@
-    value of integer/float array that is being casted. Previously the casting
-    was allowed even if the result was truncated.
+    Notes
+    -----
```
Bad merge here, this heading appears twice

mhvk commented:
@seberg, @eric-wieser, thanks for all the comments!

On the casting, I'm a bit torn. I agree with @eric-wieser that we should try to make `can_cast` consistent with actual casting behaviour in, e.g., `astype`, to the extent possible. The current behaviour of `astype`  and `__setitem__` (see below) can be summarized as follows:

- Structured to simple: only allowed for a single field and when `casting="unsafe"`; `can_cast` is inconsistent with this in always returning `True` for "unsafe", independent of the number of fields. This is an easy fix.
- Simple to structured: always done for any casting but "equiv" and "no"; this is the bug I hoped to fix. 

My own sense is that symmetry suggests that we do not allow simple to structured at all for any casting but "unsafe". But for "unsafe", it does seem fine to just allow the cast to all fields, especially as this is what happens also in `__setitem__`.

This seems to make sufficient sense that I'll go ahead and change it accordingly...

Current behaviour
-------------------------
Going from structured to simple:
```
import numpy as np
# Multiple fields to simple
i8 = np.array([1000])
i1i1 = np.array([(1,2)], dtype='i1,i1')
np.can_cast('i1,i1', 'i8', casting='unsafe')
# True # and False for any other casting mode
i1i1.astype('i8', casting='unsafe')
# ValueError: Can't cast from structure to non-structure, except if the structure only has a single field.
i8[...] = i1i1
# ValueError: Can't cast from structure to non-structure, except if the structure only has a single field.

# Single field to simple
i1 = np.array([(1,)], dtype=[('i', 'i1')])
np.can_cast([('i', 'i1')], 'i8', casting='unsafe')
# True  # and False for any other casting mode
i1.astype('i8', casting='unsafe')
# array([1])  # note that casting='unsafe' is default
i8[...] = i1
# Passes -> i8=1
np.array([(1,)], dtype=[('i', 'i1')]).astype('i8', casting='safe')
# TypeError: Cannot cast array from dtype([('i', 'i1')]) to dtype('int64') according to the rule 'safe'
```
So, for this case, `astype` behaves sensibly in allowing a cast from structured to simple only when there is a single field, and even then only when `casting="unsafe"` - the current PR would extend this to allow any case where the single field can be cast to the simple type, but I think I should revert that and instead just make `can_cast` consistent with `astype`, i.e., change its behaviour for "unsafe" to not allow the case of multiple fields.

Now turning to casting from simple to structured
```
import numpy as np
i8 = np.array([1000])
# Simple to multiple, non-safe fields
np.can_cast('i8', 'i1,i1', casting='safe')
# True  # only 'no' and 'equiv' give False
i8.astype('i1,i1', casting='safe')
# array([(-24, -24)], dtype=[('f0', 'i1'), ('f1', 'i1')])
i1i1[...] = i8
# Passes -> same array.
# only 'no' and 'equiv' error
# Behaviour identical for safe fields and for a single field!
```
So, the current behaviour is that casting to a structured type is never allowed for "equiv" and always allowed for "safe" and beyond.

mhvk commented:
p.s. I think there actually is another inconsistency for structured to structured casting, in that `can_cast' looks at field names while `astype` and `__setitem__` go by field order. Probably a leftover of #6053, so cc @ahaldane (and let's leave it out of this PR...)

mhvk commented:
Rewritten as noted above. Quite a bit simpler PR, which suggests it may be better..

mhvk commented on numpy/core/_add_newdocs.py:
```diff
@@ -2690,10 +2690,24 @@
+
+    .. versionchanged:: 1.18.0
+
+    For casting from numeric to string types, can_cast returns False
```
reworded

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -679,15 +679,62 @@ NPY_NO_EXPORT npy_bool
+    /*
+     * Fast path for basic types.
+     */
+    if ((NPY_LIKELY(from->type_num < NPY_OBJECT) &&
```
I suppose `<=` would work as well, not that it matters.

mhvk replied:
This was there before, but you're right, no sense in not including object.

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -679,15 +679,62 @@ NPY_NO_EXPORT npy_bool
+         */
+        if (!PyDataType_HASFIELDS(to) && to->type_num != NPY_OBJECT) {
+            return (casting == NPY_UNSAFE_CASTING &&
+                    PyDict_Size(from->fields) == 1);
```
I wonder if Eric won't come around a corner here and spook you with a dtype with nested fields :), it seems probable that there needs to be a recursion somewhere to get it right.

mhvk replied:
You're right, this should be recursive.

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -679,15 +679,62 @@ NPY_NO_EXPORT npy_bool
+            return 1;
+        }
+    }
+    else if (PyDataType_HASFIELDS(to)) {
```
What about object and subarrays? E.g `np.dtype("(1,2)u8")`

seberg replied:
Sorry, of course rather something like> `np.dtype(["f1", "f4", (5,)])` or `np.dtype([("f1", "f,f")])`

mhvk replied:
I think for casting to structured array, sub-arrays do not matter - certainly- `astype` happily does it. This particular stanza is really mostly there to ensure we return `False` when casting is not "unsafe".

mhvk replied:
For subarrays in `from`, `astype` seems quite happy to just take the first element. So, for now just mimicking that.

eric-wieser replied:
What happens for:
```
from_dt = np.dtype([('a', ('i4,i4', 2))])
to_dt = np.dtype([('a', 'i4')])
```

For instance
```
arr = np.array(([(1, 2), (3, 4)],), dtype=dt)
arr.astype(to_dt)
```

What's currently implemented is not "just take the first element", but "take the first element and ignore all casting rules when casting it"

mhvk replied:
Your particular example with `astype` raises an error in both in current master and in my PR, because `from_dt['a']` itself has two fields, which cannot be cast to a single one.

Sadly, though, `can_cast` happily claims this works even in my PR as `from['a']` has no fields (but does have a `subdtype`) - though only for `casting='unsafe'`.

mhvk replied:
OK, that particular case is solved. But really `astype` is weird - it checks `can_cast` internally but can still throw an error later. This should not be allowed...

seberg replied:
Well, astype casting could fail based on the values. Of course you are right and it should not fail based on information already inside the dtype.

I have to catch up a bit here, sorry, but I do not really want crunch my mind too much about it right now, we are talking about corner cases and if we have new infrastructure things _might_ get easier to align.

mhvk replied:
@seberg - agreed that we should take this bit-by-bit. My main goal really is to ensure that structured with multiple fields to unstructured always returns `False`, as this cannot possibly work (and has caused me real head-aches).

seberg reviewed (commented):
This may get more complicating,  fear... I am willing to move this forward though as an improvement though if you feel that is better. There is no regression here in 1.17, right?

mhvk commented:
OK, pushed another update. Am still trying to keep the impact minimal. Another worry I have is whether user types can have `fields` - though at least this should be better than it was before...

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -2690,10 +2690,17 @@
+    for "unsafe" casting.  Casting to multiple fields is allowed, but casting
+    from multiple fields is not.
+
+    .. versionchanged:: 1.18.0
```
I think this might be clearer as:
```
.. versionchanged:: 1.18.0
   Casting between a simple data type and ...

.. versionchanged:: 1.9.0
   Casting rom numeric to ...

mhvk replied:
Will do this once we have converged on an implementation...

mhvk replied:
I changed `versionchanged` as recommended - I think this text at least is OK.

mhvk commented:
Damn, `astype` is really quite inconsistent too in what it does for subarrays. For now, focussing on getting the structured `from` right...

charris commented:
Also failing doctest at `numpy.doc.structured_arrays`.

mhvk commented:
@seberg, @eric-wieser - what do you think we should do with this PR? Overall, I feel it at least fixes one obvious glaring inconsistency. On the other, it leaves many others. But those generally need more thought at developing a better set of tests. Personally, I'd prefer to approach this incrementally (mostly as I don't think I've got the time to do the larger overhaul...).

seberg commented:
@mhvk sorry, meant to look through this for a bit now, but then did not when I saw we had put 1.18, and thinking I can't merge it anyway ;).
I hope more restructuring of casting can/will happen, but I am ok with doing some small improvements before the end. Plus anything that works and is tested now, is something we can't break when we try to do larger changes.

mhvk commented:
Ideally this would still go in 1.17...

charris commented:
The 1.18 just means it isn't a blocker for 1.17, it can certainly go in 1.17.

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -679,15 +679,82 @@ NPY_NO_EXPORT npy_bool
+        ((NPY_LIKELY(PyDataType_ISNUMBER(from)) ||
+          PyDataType_ISOBJECT(from)) &&
+         NPY_LIKELY(from->type_num == to->type_num) &&
+         NPY_LIKELY(from->byteorder == to->byteorder))) {
```
NIT: More indentation is nice, but it really doesn't matter here.

mhvk replied:
Leaving this for the great C code cleanup...

seberg commented on numpy/core/multiarray.py:
```diff
@@ -496,11 +496,15 @@ def can_cast(from_, to, casting=None):
-    length is not long enough to store the max integer/float value converted
-    to a string. Previously can_cast in 'safe' mode returned True for
-    integer/float dtype and a string dtype of any length.
+    .. versionchanged:: 1.18.0
```
```suggestion
    .. versionchanged:: 1.17.0
```

seberg commented on numpy/core/_add_newdocs.py:
```diff
@@ -2699,10 +2699,15 @@
-    dtype to cast to is not long enough in 'safe' casting mode to hold the max
-    value of integer/float array that is being casted. Previously the casting
-    was allowed even if the result was truncated.
+    .. versionchanged:: 1.18.0
```
```suggestion
    .. versionchanged:: 1.17.0
```

seberg commented on numpy/core/tests/test_numeric.py:
```diff
@@ -899,6 +899,41 @@ def test_can_cast(self):
+        assert_(not np.can_cast('i2', [('f1', 'i4')]))
+        assert_(not np.can_cast('i2', [('f1', 'i4')], casting='same_kind'))
+        assert_(np.can_cast('i2', [('f1', 'i4')], casting='unsafe'))
+        # It should be the same for recursive structureed or subarrays.
```
```suggestion
        # It should be the same for recursive structured or subarrays.
```

seberg reviewed (approved):
OK, sorry for that. Can't find anything to squirm about right now. But I do think we should add release notes, since casting is changed in subtle ways. As most actual casting uses "unsafe" mode, I doubt that those changes affect a lot of code though (and are not already buggy).

mhvk commented:
OK, made the changes, and added a release note.

seberg commented on doc/release/1.17.0-notes.rst:
```diff
@@ -134,6 +134,12 @@ input.
+Previously, ``can_cast`` returned `True` for almost all inputs for
+``casting='unsafe'``, even for cases where casting was not possible, such as
+from a structured dtype to a regular one.  This has been fixed, making it
+more consistent with actual casting using, e.g., the ``.astype`` method.
```
There is also the change that `np.zeros(3, "i4").astype("i4,i4", casting="safe")` (or related `can_cast`) should fail after this? But maybe it is very fringe, since astype defaults to "unsafe" in any case, and so does assignment I think.

seberg replied:
Anyway, I am good with merging this.

mhvk replied:
Thanks, let me just merge this...

seberg commented:
Cool, thanks Marten.
